File: UD-04/P02-Encordados-MongoDB-JPA-Java/src/main/java/es/joseluisgs/p02encordadosmongodbjpa/models/Tenista.java
package es.joseluisgs.p02encordadosmongodbjpa.models;

import es.joseluisgs.p02encordadosmongodbjpa.enums.ManoDominante;
import es.joseluisgs.p02encordadosmongodbjpa.enums.ManoDominanteUtils;
import es.joseluisgs.p02encordadosmongodbjpa.enums.TipoReves;
import es.joseluisgs.p02encordadosmongodbjpa.enums.TipoRevesUtils;
import es.joseluisgs.p02encordadosmongodbjpa.utils.MyLocale;
import org.bson.types.ObjectId;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.DocumentReference;

import java.time.LocalDate;
import java.util.Objects;


@Document("tenistas")
public class Tenista {
    @Id
    private ObjectId id;
    private String nombre;
    private Integer ranking;
    private LocalDate fechaNacimiento;
    private Integer añoProfesional;
    private Integer altura;
    private Integer peso;
    private Double ganancias;
    private ManoDominante manoDominante;
    private TipoReves tipoReves;
    private Integer puntos;
    private String pais;
    // Mi referencia a la raqueta

    // Apuntamos a la raqueta por su id en modo referencia M-1
    //@DocumentReference(lazy = true)
    @DocumentReference
    private Raqueta raqueta;

    public Tenista(String nombre, Integer ranking, LocalDate fechaNacimiento, Integer añoProfesional, Integer altura,
                   Integer peso, Double ganancias, ManoDominante manoDominante, TipoReves tipoReves,
                   Integer puntos, String pais) {
        this.nombre = nombre;
        this.ranking = ranking;
        this.fechaNacimiento = fechaNacimiento;
        this.añoProfesional = añoProfesional;
        this.altura = altura;
        this.peso = peso;
        this.ganancias = ganancias;
        this.manoDominante = manoDominante;
        this.tipoReves = tipoReves;
        this.puntos = puntos;
        this.pais = pais;
    }

    public Tenista() {
    }

    public ObjectId getId() {
        return id;
    }

    private void setId(ObjectId id) {
        this.id = id;
    }


    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public Integer getRanking() {
        return ranking;
    }

    public void setRanking(Integer ranking) {
        this.ranking = ranking;
    }

    public LocalDate getFechaNacimiento() {
        return fechaNacimiento;
    }

    public void setFechaNacimiento(LocalDate fechaNacimiento) {
        this.fechaNacimiento = fechaNacimiento;
    }

    public Integer getAñoProfesional() {
        return añoProfesional;
    }

    public void setAñoProfesional(Integer añoProfesional) {
        this.añoProfesional = añoProfesional;
    }

    public Integer getAltura() {
        return altura;
    }

    public void setAltura(Integer altura) {
        this.altura = altura;
    }

    public Integer getPeso() {
        return peso;
    }

    public void setPeso(Integer peso) {
        this.peso = peso;
    }

    public Double getGanancias() {
        return ganancias;
    }

    public void setGanancias(Double ganancias) {
        this.ganancias = ganancias;
    }

    public ManoDominante getManoDominante() {
        return manoDominante;
    }

    public void setManoDominante(ManoDominante manoDominante) {
        this.manoDominante = manoDominante;
    }

    public TipoReves getTipoReves() {
        return tipoReves;
    }

    public void setTipoReves(TipoReves tipoReves) {
        this.tipoReves = tipoReves;
    }

    public Integer getPuntos() {
        return puntos;
    }

    public void setPuntos(Integer puntos) {
        this.puntos = puntos;
    }

    public String getPais() {
        return pais;
    }

    public void setPais(String pais) {
        this.pais = pais;
    }


    public Raqueta getRaqueta() {
        return raqueta;
    }

    public void setRaqueta(Raqueta raqueta) {
        this.raqueta = raqueta;
    }


    public void removeRaqueta() {
        this.raqueta.getTenistas().remove(this);
        this.raqueta = null;
    }

    @Override
    public String toString() {
        return "Tenista{" +
                "uuid=" + id +
                ", nombre='" + nombre + '\'' +
                ", ranking=" + ranking +
                ", fechaNacimiento=" + MyLocale.toLocalDate(fechaNacimiento) +
                ", añoProfesional=" + añoProfesional +
                ", altura=" + altura +
                ", peso=" + peso +
                ", ganancias=" + MyLocale.toLocalMoney(ganancias) +
                ", manoDominante=" + ManoDominanteUtils.toString(manoDominante) +
                ", tipoReves=" + TipoRevesUtils.toString(tipoReves) +
                ", puntos=" + puntos +
                ", pais='" + pais + '\'' +
                ", raqueta=" + raqueta +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Tenista tenista)) return false;
        return id.equals(tenista.id) && nombre.equals(tenista.nombre) && ranking.equals(tenista.ranking) && fechaNacimiento.equals(tenista.fechaNacimiento) && añoProfesional.equals(tenista.añoProfesional) && altura.equals(tenista.altura) && peso.equals(tenista.peso) && ganancias.equals(tenista.ganancias) && manoDominante == tenista.manoDominante && tipoReves == tenista.tipoReves && puntos.equals(tenista.puntos) && pais.equals(tenista.pais);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, nombre, ranking, fechaNacimiento, añoProfesional, altura, peso, ganancias, manoDominante, tipoReves, puntos, pais);
    }
}

File: UD-04/P03-Encordados-MongoDB-JPA-Kotlin/src/main/kotlin/es/joseluisgs/p02encordadosmongodbjpakotlin/models/Tenista.kt
package es.joseluisgs.p02encordadosmongodbjpakotlin.models

import es.joseluisgs.p02encordadosmongodbjpakotlin.extensions.toLocalDate
import es.joseluisgs.p02encordadosmongodbjpakotlin.extensions.toLocalMoney
import es.joseluisgs.p02encordadosmongodbjpakotlin.extensions.toLocalNumber
import org.bson.types.ObjectId
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import org.springframework.data.mongodb.core.mapping.DocumentReference
import java.time.LocalDate
import java.util.*


@Document("tenistas")
data class Tenista(
    @Id
    val id: ObjectId = ObjectId.get(),
    var nombre: String,
    var ranking: Int,
    var fechaNacimiento: LocalDate,
    var añoProfesional: Int,
    var altura: Int,
    var peso: Int,
    var ganancias: Double,
    var manoDominante: ManoDominante,
    var tipoReves: TipoReves,
    var puntos: Int,
    var pais: String,

    // Mi referencia a la raqueta
    // Apuntamos a la raqueta por su id en modo referencia M-1
    @DocumentReference()
    //@DocumentReference(lookup = "{ '_id' : ?#{#target} }") // Otra forma de ponerlos
    var raqueta: Raqueta? = null

) {

    override fun toString(): String {
        return "Tenista(uuid=$id, nombre='$nombre', ranking=$ranking, " +
                "fechaNacimiento=${fechaNacimiento.toLocalDate()}, " +
                "añoProfesional=$añoProfesional, " +
                "altura=${(altura.toDouble() / 100).toLocalNumber()} cm, " +
                "peso=$peso Kg, " +
                "ganancias=${ganancias.toLocalMoney()}, " +
                "manoDominante=${manoDominante.mano}, " +
                "tipoReves=${tipoReves.tipo}, " +
                "puntos=$puntos, pais=$pais, " +
                // Cuidado con la recursividad, porque al imprimir la raqueta, esta a su vez imprime el tenista y
                // así sucesivamente entramos en un bucle infinito
                "raqueta=${raqueta})"
    }

    // ENUMS de la propia clase
    enum class ManoDominante(val mano: String) {
        DERECHA("DERECHA"),
        IZQUIERDA("IZQUIERDA");

        companion object {
            fun from(manoDominante: String): ManoDominante {
                return when (manoDominante.uppercase()) {
                    "DERECHA" -> DERECHA
                    "IZQUIERDA" -> IZQUIERDA
                    else -> throw IllegalArgumentException("ManoDominante no reconocida")
                }
            }
        }
    }

    enum class TipoReves(val tipo: String) {
        UNA_MANO("UNA MANO"),
        DOS_MANOS("DOS MANOS");

        companion object {
            fun from(tipoReves: String): TipoReves {
                return when (tipoReves.uppercase()) {
                    "UNA MANO" -> UNA_MANO
                    "DOS MANOS" -> DOS_MANOS
                    else -> throw IllegalArgumentException("TipoReves no reconocida")
                }
            }
        }
    }


}